### ISSUE TITLE
Custom fall damage

### ISSUE THREAD

**Issue by Wuzzy2** [luanti-org/luanti#10348]
## Problem

Every aspect of fall damage is hardcoded: The fact that you take fall damage in the first place, how much fall damage you take, at which speed/height you start to take fall damage, the formula that calculates fall damage, etc.

This is bad as it limits games quite a bit.

## Solutions

Make fall damage customizable. The bare minimum should be a method to disable fall damage altogether (while keeping the damage system).
Disabling fall damage could be done with a player physic (`set_physics_override`) or a special armor group.

The trickier part is to decide about what should be done for customizing fall damage in detail.

There are several aspects of fall damage:
1) Scaling variables to increase or decrease the vulnerability to fall damage. E.g. do you want to make players take damage at 4 nodes? 5 nodes? 6 nodes? Etc.
2) Which inputs you use to calculate fall damage. I can think of two possible inputs that might be used for a calculation: Your impact speed, or the height from which you fell from. I believe, currently, Minetest only cares about the first one.
3) The formula that calculates fall damage

Number 1 might be the easiest and least disruptive one to do since nothing major about the fall damage system is changed. Fall damage could be made subject to a special armor group `fall_damage`. If `0` or `nil`, no change, but if `50`, fall damage is 50% increased.
Number 2 is more tricky to do. As for the “fall height” based fall damage, since there is no formula for this type of fall damage, one needs to be invented first.
Number 3 will be the most tricky to do. The question is how to teach the client this new formula. One solution might be use a string and some kind of mini-language that supports minimal mathematical operations and this is sent to clients. I don't know how many operators you would need, but I believe in practice, it's never going to be more than addition, subtraction, division, multiplication and exponentiation.
Another solution might be to wait until client scripting is a thing, but this comes with its own host of problems.

My solutions are not set in stone, I believe there are a billion other ways to tackle this issue.

Finally, I think no matter how this problem is tackled, I obviously advocate for doing so in a backwards-compatible manner. That is, existing games should have their fall damage still function like before, with no changes at all.

**Comment by hecktest** [comment]
Reasonable problem but terrible solution.

There is absolutely no reason this should live in the binary, and there is absolutely no reason the client should know or care about fall damage. I think I've mentioned this before; the client gets to "predict" something that's up to the server anyway, a whole 100ms ahead of time. Wow.
What is needed is a generic "impact" event, when an ObjectRef collides with something - a node, or another object.
The default handler for this impact event can be an emulator of the current fall damage, with a warning if you don't replace it.
By 6.0 the fallback should be gone entirely, by that time it's reasonable to expect all subgame maintainers to paste a little snippet of code, or depend on a common "falldamage" mod which can be distributed with the engine.

**Comment by paramat** [comment]
hecktest, interesting ideas, but assuming fall damage is handled as currently ...

The only correct way to determine fall damage is by velocity change on impact, as currently. Fall height is irrelevant as gravity can be changed, very gently floating down from 6 nodes on a dwarf planet should not cause damage.
So forget about fall distance, and doing this simplifies the consideration.

The velocity change threshold for damage could be made settable, setting it to a very high value would then disable fall damage, so you would get both features from only one value applied using set_physics_override().
Damage done above the threshold should be calculated as currently for consistency, there is no reason to change that. That would also ensure backwards compatibility.
Custom fall damage formulas do not make sense in terms of physics, and are unnecessarily overcomplex.

So, only one setting value is needed, that is all i would support.

**Comment by Wuzzy2** [comment]
Using fall height as input value can still be useful, purely for gameplay reasons. You can use this to increase fall damage linearly instead of exponentially. This can give very different gameplay mechanics. Remember that Minetest is not a realistic physics simulation, so dismissing this option because of realism is annoying. There is no single “correct” way to calculate fall damage. But obvious downside is of course increased complexity …



> There is absolutely no reason this should live in the binary, and there is absolutely no reason the client should know or care about fall damage. I think I've mentioned this before; the client gets to "predict" something that's up to the server anyway, a whole 100ms ahead of time. Wow.

100ms is actually long enough to be noticable by the player. Take away client prediction and you WILL instantly see the difference. Client prediction is actually a good thing. The problem is not that the client predicts something, the problem is when the client predicts something wrong. The solution to that would be to teach the client what to predict.

> What is needed is a generic "impact" event, when an ObjectRef collides with something - a node, or another object.
The default handler for this impact event can be an emulator of the current fall damage, with a warning if you don't replace it.

But then how do you do client prediction? I think this is quite important.

> By 6.0 the fallback should be gone entirely, by that time it's reasonable to expect all subgame maintainers to paste a little snippet of code, or depend on a common "falldamage" mod which can be distributed with the engine.

Removing functionality and intentionally breaking games just for the sake of it should not be done. Fall damage is so ingrained in so many games, removing it would be a massive middle finger to all game authors. And there's no really good reason to do so anyway, it's not like the fallback would be ultra complex or anything.

TBH, I am siding with paramat here by sticking with simple solutions. Simple scaling values will already help so much.

----
By the way, I have found the code that calculates fall damage:

```
f32 pre_factor = 1; // 1 hp per node/s
f32 tolerance = BS*14; // 5 without damage
f32 post_factor = 1; // 1 hp per node/s
if (info.type == COLLISION_NODE) {
	const ContentFeatures &f = m_client->ndef()->
		get(m_map->getNode(info.node_p));
	// Determine fall damage multiplier
	int addp = itemgroup_get(f.groups, "fall_damage_add_percent");
	pre_factor = 1.0f + (float)addp / 100.0f;
}
float speed = pre_factor * speed_diff.getLength();
if (speed > tolerance && !player_immortal) {
	f32 damage_f = (speed - tolerance) / BS * post_factor;
	u16 damage = (u16)MYMIN(damage_f + 0.5, U16_MAX);
	// Deal damage ...
	// ...
}
```

So, up to a minimal vertical Y speed (`tolerance`), no fall damage is given. When the minimal speed is passed, fall damage increases linearly with speed.

So if we're going for the simple solution of only supporting scaling values, a scaling value like paramat suggested is obviously a good idea. But you would obviously also want a changable `tolerance` to determine how fast you have to fall at minimum before you start taking damage.

**Comment by sorcerykid** [comment]
I think we really need to forego fall damage, and provide a means to handle custom impact damage instead. 

In my Mobs Lite engine, for example I check for impact along _any_ axis and then apply a damage formula accordingly. This works very well for entities. Now if something similar were possible for players, that would be excellent. If anything, this seems like yet another use-case for server-sent CSM.

**Comment by sfan5** [comment]
One issue with that would be that SSCSM is supposed to be optional, but physics handling is entirely client-side.

**Comment by paramat** [comment]
This feature is unsuitable for SSCSM because it is a fundamental physics customisation like custom speed, jump, gravity etc. Something this fundamental should not rely on SSCSM, it should be part of per-player physics override.
(Besides, SSCSM is 3 years overdue and is not progressing, it is uncertain it will ever happen).
Anyway, assuming this is done simply as i described, with 1 value set in physics override, i support the feature :+1: 

Wuzzy2,
> Using fall height as input value can still be useful, purely for gameplay reasons

It is simply impossible to use 'fall height', as we cannot track how far a player falls. Besides, as i explained, fall height is irrelevant to fall damage due to gravity being variable. Impact speed is the only sane method.
> Remember that Minetest is not a realistic physics simulation, so dismissing this option because of realism is annoying

It is not a 'highly realistic physics simulation' correct, but that does not mean it supports all possible wacky alternative universe physics, it does not for simplicity and because wacky physics is very rarely useful and should be programmed in Lua by those very few who need it, to not bloat the engine with mostly useless features.
MTEngine physics is indeed simplified realistic physics.
> So if we're going for the simple solution of only supporting scaling values, a scaling value like paramat suggested is obviously a good idea. But you would obviously also want a changable tolerance

I am suggesting variable tolerance only, not variable damage scaling. There is very little or zero usefulness to users in variable scaling above the tolerance, that is overkill customisation.

However, i propose a change in the scaling above the tolerance to make sense:

Currently, there is 14 node/s of tolerance, and 34 node/s kills. Very roughly, twice the tolerance kills.

Now imagine, for sake of argument, the tolerance is set to 1000 node/s. 1020 node/s would kill, which is silly because it is such a tiny increase over 1000, it should be more like 2000 node/s to kill.
I can share the exact math later, but i suggest a scaling of damage above the tolerance that is proportional to tolerance, such that roughly twice tolerance kills, always.
The maths would also be designed to give current damage scaling behaviour when tolerance is set to the current 14 node/s, for backwards compatibility.

I am offering to code the damage calculation, i could probably do the whole feature. Let me assign myself.

**Comment by sorcerykid** [comment]
Then the new calculation should be based on speed of impact in any direction, as has already been requested.

https://forum.minetest.net/viewtopic.php?f=5&t=23799

**Comment by Wuzzy2** [comment]
Yeah, I am fine with calculating based on velocity alone. The other calculation based on fall height was only an esoteric use case anyway, no big deal.

**Comment by paramat** [comment]
> Then the new calculation should be based on speed of impact in any direction

Horizontal impact damage is intentionally disabled to avoid damage when using 'fast move' or very fast sprint potions, according to comments in the code.
We could make all-direction damage optional, that seems fine to me, but that should be a separate PR.
Lack of damage when moving upwards was discussed in #8051 , there is discussion there about what directions could cause damage.
There seems ot be no feature request for optional all-direction damage, maybe someone would like to open one?

**Comment by hecktest** [comment]
No, none of this requires SSCSM. Where did you people get that idea? Damage is not a client thing.

Also, omnidirectional impact damage is just a somewhat crappy version of the collision event I proposed.
_Why am I even arguing this? I already have custom fall damage and collision, this is only a problem for the poor victims of the hardcoded health system._

**Comment by sorcerykid** [comment]
> No, none of this requires SSCSM. Where did you people get that idea? Damage is not a client thing.

Damage of players is calculated on the server? I thought the server was only responsible for collision detection of LuaEntitySAOs.

**Comment by paramat** [comment]
I wondered about this too and checked the code, fall damage is calculated by the client, could not find any serverside calculation (that would require serverside player movement modelling).

**Comment by hecktest** [comment]
Let me correct: damage _has no business_ being a client's task. If you expect every client to honestly report damage, you'll have a lot of cheaters on your hands.

**Comment by sorcerykid** [comment]
Agreed, it does seems counter-intuitive that the damage calculation is the sole responsibility of the client. I get the distinct impression that the entire game mechanic needs to be re-evaluated, because as it currently stands, there are a lot of fundamental limitations of the current system.

**Comment by appgurueu** [comment]
Players should provide a moveresult like entities already do. The server (overridable builtin luamethod `minetest.calculate_fall_damage(player, moveresult)`) should calculate fall damage based on this. Prediction can be kept but needs to be customizable.

**Comment by TheTermos** [comment]
> Let me correct: damage _has no business_ being a client's task.

More general, giving the client any authority over what happens to player is a questionable design, but as of now it has all of it. That creates a whole bunch of unsolvable problems, besides non-moddability and hackability there's also that players can't be synchronized with anything, e.g. #9806 where tumbleweed rolls in the wind.

> As movement currently is clientside, collisions would be provided by the client though, still allowing cheaters to take no fall damage.

There are even more problems. Moveresult is only valid during a single move, but client and server steps are independent, and player makes multiple moves per client step.
With dedicated server the step duration is by default minimum 9 times longer than player move, without even accounting for any delays.
Bottom line you can try to have the client send collision info to the server, but it's guaranteed it's going to be always wrong.

**Comment by appgurueu** [comment]
> Bottom line you can try to have the client send collision info to the server, but it's guaranteed it's going to be always wrong.

Wrong but probably good enough, and certainly better than what we currently have.

**Comment by mk-pmb** [comment]
> Now imagine, for sake of argument, the tolerance is set to 1000 node/s. 1020 node/s would kill, which is silly because it is such a tiny increase over 1000,

I don't think so. There are many potential scenarios where a sharp non-linear threshold occurrs in physics. One most easily imagined is bending a spaghetti. Up to a certain curvature, the changes are almost entirely reversible, no lasting damage.

**Comment by sorcerykid** [comment]
Perhaps we could look at how other multiplayer games handle this? Like how did the original Quake by ID Software solve these types of problems? Was player movement solely managed client-side, or did the server have an integral role?

**Comment by mk-pmb** [comment]
@sorcerykid What lack of information do you see that such research could help us with? Maybe I missed some part.

**Comment by sorcerykid** [comment]
> What lack of information do you see that such research could help us with? Maybe I missed some part.

I presume that you're being facetious, or else you haven't read through this issue at all.

**Comment by appgurueu** [comment]
A fall damage armor group combined with a serverside player collision handler (with the client sending collisions to the player) could come in handy. In order to reduce the power of client modifications, ideally the server should run it's own player movement code and somehow verify whether the results are similar. This is, however, nontrivial.

**Comment by TheTermos** [comment]
> Perhaps we could look at how other multiplayer games handle this? Like how did the original Quake by ID Software solve these types of problems?

I think MT should be considered according to it's own merits.
Original Quake was primarily a single player game with multiplayer mode added on top iirc, while MT uses client-server model even for single player mode.
Also Quake has fast paced gameplay where split second reflexes matter, hence quite different requirements. And there are many more important differences.

Still here's an article I've found about the network model of Quake 3 which is actually a multiplayer game.
https://fabiensanglard.net/quake3/network.php
MT isn't nearly as reflexes based so the snapshot system as robust as described therein wouldn't probably be necessary.

On the other hand Unity seems to give the client the authority over player objects by default, but so what.
I played a bit of Battlestar Galactica Online made with Unity, and problems created by that model were very apparent, e.g  weapon fire rate was dependent on client fps. That was probably one of the factors that sealed the game's fate.

Anyway this particular issue is about fall damage, for discussing the entirety of network model a separate issue would be in order. The staff makes the impression of trying hard to avoid discussing this problem.

**Comment by Wuzzy2** [comment]
I am now convinced that ripping out fall damage handling from the client is a precondition for this issue. As long as fall damage lives in the client, we can't even begin on customization options.

#11025

**Comment by Wuzzy2** [comment]
I think this issue is technically fixed, and it has been for a long time.
Now, the hardcoded fall damage can be completely disabled per-player by giving them the armor group `add_fall_damage_percent=-100`. If this code is added, games and mods can now roll their own fall damage system, if they so desire.

Closing.
